add arg support for Int16, Int8, UInt16, UInt8

baseballlover723 opened the pull request:
I was trying to use Jennifer with sqlite with an Int16 in my model, and I got an error saying that it was supported. Upon looking into it, it seems that sqlite only stores regular Int32s (and it works if I switch the type to Int32). However, given that Int8, Int16, UInt8, and UInt16 can all fit in a Int32. I don't see why we can't just convert them to Int32's. There is already a similar thing happening for Float32 being converted to Float64, and this is essentially the same thing but for ints.

I didn't see any tests for any of the other binds, so I didn't add any here.

bcardiff commented:
ResultSet should be able to work with the same types as Statement. Would you add the overloads?

I don't have concern on signed ints being added. But adding unsigned ones goes against the idea of using them mostly for bit/protocol implementation rather than domain logic. I'm not against adding them, but they feel odd. Why don't put them in crystal-db directly to enforce them on every driver?

baseballlover723 commented:
> ResultSet should be able to work with the same types as Statement. Would you add the overloads?

I have added the overloads

> I don't have concern on signed ints being added. But adding unsigned ones goes against the idea of using them mostly for bit/protocol implementation rather than domain logic. I'm not against adding them, but they feel odd. Why don't put them in crystal-db directly to enforce them on every driver?

I've removed the unsigned int parts, since I'm not actually using unsigned ints and I don't have much bandwidth. I mostly just added them because to it seemed to be the same as the lower int ones, which it seems they aren't / shouldn't be.

baseballlover723 commented:
> ResultSet should be able to work with the same types as Statement. Would you add the overloads?

I have added the overloads

> I don't have concern on signed ints being added. But adding unsigned ones goes against the idea of using them mostly for bit/protocol implementation rather than domain logic. I'm not against adding them, but they feel odd. Why don't put them in crystal-db directly to enforce them on every driver?

I'd prefer to have it here, at least for now (though crystal-db may be the better spot for it), because I'm really sure that this conversion can be made that generally. Other databases with multiple int types may want to handle things differently imo and I don't want to drive something that I don't have much experience in, nor much time to dive any deeper into this. And to me, this seems to be sqlite specific, as sqlite only has 1 integer type (and other dbs with more integer types may want to do other things). And I would expect that if sqlite can work with a larger int type, then it should be able to work with smaller unsigned int types that can be fully encapsulated in the larger int type.

bcardiff reviewed (approved):
Thanks for addressing the requests so quickly

bcardiff commented on src/sqlite3/statement.cr:
```diff
@@ -45,6 +45,26 @@ class SQLite3::Statement < DB::Statement
+  end
+
+  private def bind_arg(index, value : UInt32)
+    check LibSQLite3.bind_int(self, index, value)
```
Shouldn't this use int64 conversion? Otherwise is casting it to int32 and the highest values would be stored as negatives i believe.

If bandwidth is over i'll try to add some tests to check for this.

baseballlover723 replied:
you are correct (though I didn't test it, since I'm only using UInt16 in my application). I just copy pasted it from the int ones and forgot that UInt32 needs to go into the 64 bit Int unlike the Int32 one. I've pushed a change to fix it.

As for tests, I unfortunately don't have bandwidth anymore since labor day is now over. I took a brief look at adding some more sample values in db_spec, but the complier complained about the types, so it seems like that'll have to tie in with crystal-db somewhere, but I think you all are probably better suited for making changes there. At the very least, my setup using Jennifer is now working with UInt16 and Int16. So I'd expect it be sound.

bcardiff commented:
Hmm. for some reason I can't push to your branch. I will open another PR with the specs.

Note: UInt64 is not supported because sqlite integer types are always signed up to 8 bytes. So trying to store a UInt64 will not fit.
